chain/client/src/archive/cloud_archival_actor.rs

```diff
@@ -0,0 +1,173 @@
+    Ok(Some((actor, keep_going_clone)))
+}
+
+impl Actor for CloudArchivalActor {
+    fn start_actor(&mut self, ctx: &mut dyn DelayedActionRunner<Self>) {
+        tracing::info!(target: "cloud_archival", "Starting the cloud archival actor");
+        self.cloud_archival_loop(ctx);
+    }
+}
+
+impl CloudArchivalActor {
+    /// Main loop: archive as fast as possible until `cloud_head == hot_final_head`, then sleep for
+    /// `polling_interval` before trying again.
+    fn cloud_archival_loop(&mut self, ctx: &mut dyn DelayedActionRunner<Self>) {
+        if !self.keep_going.load(std::sync::atomic::Ordering::Relaxed) {
+            tracing::debug!(target: "cloud_archival", "Stopping the cloud archival loop");
+            return;
+        }
+
+        let result = self.try_archive_data();
+
+        let duration = if let Ok(CloudArchivingResult::OlderHeightArchived) = result {
+            Duration::ZERO
+        } else {
+            self.config.polling_interval
+        };
+
+        ctx.run_later("cloud_archival_loop", duration, move |actor, ctx| {
+            actor.cloud_archival_loop(ctx);
+        });
+    }
+
+    /// Tries to archive one height and logs the outcome.
+    fn try_archive_data(&mut self) -> anyhow::Result<CloudArchivingResult, CloudArchivingError> {
+        // TODO(cloud_archival) Add metrics
+        let result = self.try_archive_data_impl();
+
+        match &result {
+            Err(err) => {
+                tracing::error!(target : "cloud_archival", error = format!("{err:#?}"), "Archiving data to cloud failed");
+            }
+            Ok(archive_result) => match archive_result {
+                CloudArchivingResult::NoHeightArchived => {
+                    tracing::trace!(target: "cloud_archival", "No height was archived - cloud archival head is up to date");
+                }
+                CloudArchivingResult::LatestHeightArchived => {
+                    tracing::trace!(target: "cloud_archival", "Latest height was archived");
+                }
+                CloudArchivingResult::OlderHeightArchived => {
+                    tracing::trace!(target: "cloud_archival", "Older height was archived - more archiving needed");
+                }
+            },
+        }
+        result
+    }
+
+    /// If the cloud head lags the hot final head, archive the next height. Updates `cloud_head` on success.
+    fn try_archive_data_impl(
+        &mut self,
+    ) -> anyhow::Result<CloudArchivingResult, CloudArchivingError> {
+        let _span = tracing::debug_span!(target: "cloud_archival", "cloud_archive").entered();
+
+        let target_height = hot_final_head_height(&self.hot_store, self.genesis_height)?;
+        if self.cloud_head >= target_height {
+            return Ok(CloudArchivingResult::NoHeightArchived);
+        }
+        let height_to_archive = self.cloud_head + 1;
+        self.archive_data(height_to_archive)?;
+        self.update_cloud_archival_head(height_to_archive)?;
```

Comment:
I guess if this function fails, we risk uploading the archive data again? Do buckets support overwriting files?

Reply:
Yes, we can overwrite.